libqtile/widget/widgetbox.py

```diff
@@ -0,0 +1,144 @@
+# SOFTWARE.
+
+from libqtile import bar
+from libqtile.widget import base
+
+
+class WidgetBox(base._Widget):
+    """A widget to declutter your bar.
+
+    WidgetBox is a widget that hides widgets by default but shows them when
+    the box is opened.
+
+    Widgets that are hidden will still update etc. as if they were on the main
+    bar.
+
+    Button clicks are passed to widgets when they are visible so callbacks will
+    work.
+
+    Widgets in the box alsos remain accessible via command interfaces.
+
+    Widgets can only be added to the box via the configuration file. The widget
+    is conigured by adding widgets to the "widgets" parameter as follows:
+
+        widget.WidgetBox(widgets=[
+            widget.TextBox(text="This widget is in the box"),
+            widget.Memory(),
+            ]),
+    """
+    orientations = base.ORIENTATION_HORIZONTAL
+    defaults = [
+        ("font", "sans", "Text font"),
+        ("fontsize", None, "Font pixel size. Calculated if None."),
+        ("fontshadow", None, "font shadow color, default is None(no shadow)"),
+        ("foreground", "#ffffff", "Foreground colour."),
+        ("text_closed", "[<]", "Text when box is closed"),
+        ("text_open", "[>]", "Text when box is open"),
+    ]
+
+    def __init__(self, widgets=list(), **config):
+        base._Widget.__init__(self, bar.CALCULATED, **config)
+        self.add_defaults(WidgetBox.defaults)
+        self.box_is_open = False
+        self.widgets = widgets
+
+    def _configure(self, qtile, bar):
+        base._Widget._configure(self, qtile, bar)
+        self.set_box_icon()
+
+        for idx, i in enumerate(self.widgets):
+            if i.configured:
+                i = i.create_mirror()
+                self.widgets[idx] = i
+            qtile.register_widget(i)
+            i._configure(qtile, bar)
+
+    def calculate_length(self):
+        width = self.layout.width
+
+        if self.box_is_open and self.widgets:
+            width += sum(i.width for i in self.widgets)
+
+        return width
+
+    def set_box_icon(self):
+        self.layout = self.drawer.textlayout(
+            self.text_open if self.box_is_open else self.text_closed,
+            self.foreground,
+            self.font,
+            self.fontsize,
+            self.fontshadow,
+            markup=False,
+        )
+
+    def draw(self):
+        self.drawer.clear(self.background or self.bar.background)
+
+        if self.offsetx is None:
+            return
+        offset = self.offsetx
+
+        self.layout.draw(0,
+                         int(self.bar.height / 2.0 -
+                             self.layout.height / 2.0) + 1)
+
+        offset += self.layout.width
+
+        self.drawer.draw(offsetx=self.offsetx, width=self.width)
+
+        for i in self.widgets:
+            i.offsetx = offset
+            offset += i.width
+            i.draw()
+
+    def button_press(self, x, y, button):
```

Comment:
Can you use mouse_callbacks for this, or will that not work?

Reply:
This is just for the WidgetBox. I need the x value to work out which widget has been clicked. My thinking was that, by then calling the child's button_press method, it would still trigger the mouse_callbacks.

I don't think using mouse_callbacks for the WidgetBox itself makes sense because we just want to pass these events. I can be persuaded otherwise if you think there's a better approach.

Reply:
When you click on a child widget, does the click event not go directly to that widget rather than into the WidgetBox widget?

Reply:
Interesting question. I don't _think_ so but it's probably worth checking. 

I assumed that, given the WidgetBox's width is calculated as the total width of the widgets inside it, the bar would assume that any click in that area was just passed to the WidgetBox.

Reply:
Checked. If I remove the button handling then the widgets in the box don't register and clicks. Passing the events the way I do seems to work but I may be missing a better way?

I also haven't passed mouse_enter/leave events currently. That seems harder to implement here so probably need to include something in the docs.

Reply:
I think it is fine how you override `button_press` because it does need unique logic, but I think it should still funnel into `mouse_callbacks`. Consider something like this:
```
    def button_press(self, x, y, button):
        widget, x_pos = self.find_widget(x)

        if widget == self:
            name = 'Button{0}'.format(button)
            if name in self.mouse_callbacks:
                self.mouse_callbacks[name]()

        elif widget is not None:
            widget.button_press(x_pos, y, button)
```
Then you can get rid of the `clicked` method, and add `self.add_callbacks({'Button1': self.toggle_box})` to `__init__`. This way, more callbacks can be added, or the button1 callback can be overriden.

Reply:
As for `mouse_enter` and `mouse_leave`, you can use `find_widget` the same:
```
    def mouse_enter(self, x, y):
        widget, x_pos = self.find_widget(x)
        widget.mouse_enter(x_pos, y)
```

Reply:
@m-col : yes, that could work for `mouse_callbacks`. Happy to make that change.

As for `mouse_enter` the above idea only when you enter the widget box. The problem I have is when you're already in the box but move horizontally over different widgets in the box. In that scenario, the mouse_enter wouldn't get triggered more than once. If we can't sort that then `mouse_leave` would potentially fire in a different widget to the one that was entered.

We would constantly need to know mouse_position in the widget to implement this effectively.

Reply:
Ah yes....  Is there any way you think the child widgets can be handled in a way that they are actually part of the bar's normal widget list, rather than a 'sub'-widget? That would also simplify the button press logic because everything would be direct. Maybe to hide and show the widgets all you need to do is move them in and out of the bar's `self.widgets` list.

Reply:
That has a horrible sound of being a far better way for this widget to work! Will investigate.

Reply:
Genius. So much simpler. Thanks.

Reply:
Does that approach work completely? Widgets that are hidden will continue doing their polling or other behaviour in the background, which is fine, but some will modify their `self.drawer` directly even if they are not within a bar's widget list. Does this not cause issues? The `Drawer`s will paint onto the bar's window, no?

Reply:
I should also add that mirrored widgets are not working for me. I.e. if I do this:
```
clock = widget.Clock()
widgetbox = widget.WidgetBox(widgets=[clock])
mybar = bar.Bar([widgetbox, clock])
```
`clock` within the bar's widget list is replaced with a `Mirror` widget, copying the pixels from the `clock` widget that is held by `widgetbox`. For some reason, likely related to how drawing is being handled, the mirror widgets aren't updating unless the widgetbox is visible.

Reply:
> Does that approach work completely? Widgets that are hidden will continue doing their polling or other behaviour in the background, which is fine, but some will modify their `self.drawer` directly even if they are not within a bar's widget list. Does this not cause issues? The `Drawer`s will paint onto the bar's window, no?

Interesting. I hadn't experienced this but I can see your point. Not sure what the fix is for this because I think the widgets should continue to poll.

Reply:
> I should also add that mirrored widgets are not working for me. I.e. if I do this:
> 
> ```
> clock = widget.Clock()
> widgetbox = widget.WidgetBox(widgets=[clock])
> mybar = bar.Bar([widgetbox, clock])
> ```
> 
> `clock` within the bar's widget list is replaced with a `Mirror` widget, copying the pixels from the `clock` widget that is held by `widgetbox`. For some reason, likely related to how drawing is being handled, the mirror widgets aren't updating unless the widgetbox is visible.

I'm not familiar with mirror widgets. What's the expected behaviour here?

Reply:
> Interesting. I hadn't experienced this but I can see your point. Not sure what the fix is for this because I think the widgets should continue to poll.

Can you confirm with a bunch of different widgets whether or not this is actually a problem? The issue isn't that they continue to poll (that is to be expected), but that they will by themselves try to draw pixels onto the bar's window. A potential solution would be to hijack each widget's `draw` method and substitute it when hidden with a no-op. Then revert it to show the widget. You may or may not need to do the bar list manipulation... not sure.

> I'm not familiar with mirror widgets. What's the expected behaviour here?

You can see the implementation [here](https://github.com/qtile/qtile/blob/master/libqtile/widget/base.py#L621), and where it is used within `Bar._configure`, including the block that you've copied here. They let users use a single widget instance (and therefore only one object doing tasks) in multiple places on the same bar, or on multiple bars. Any time a widget gets `_configure`-ed a further time after the first time, it gets replaced by a mirror widget that copies the original's pixels. They get their own `Drawer` instance that hooks into the drawing of the widget that they mirror's drawer, but I suspect that no-oping their `draw` method as above will probably fix any potential issues.

Reply:
Yes, can do a lot more testing over the weekend. As for hijacking the draw function, the was the approach that I was considering so I'm glad to see I'm on the right track!

Reply:
> I should also add that mirrored widgets are not working for me. I.e. if I do this:
> 
> ```
> clock = widget.Clock()
> widgetbox = widget.WidgetBox(widgets=[clock])
> mybar = bar.Bar([widgetbox, clock])
> ```
> 
> `clock` within the bar's widget list is replaced with a `Mirror` widget, copying the pixels from the `clock` widget that is held by `widgetbox`. For some reason, likely related to how drawing is being handled, the mirror widgets aren't updating unless the widgetbox is visible.

OK - I understand what you mean now. That's an interesting issue because, based on the other issue you identified, we need to suppress the hidden widget's draw function. We could prevent mirroring by forcing the widgets in the box to have `configured = False` but that feels a bit ugly and might cause unexpected issues if widgets rely on that flag for other purposes.

Will do some more thinking about this.

Reply:
Ok... so I _think_ I've got this worked out. For the `Mirror` to work, the widget still needs to prepare its surface so we can't just detach the `draw` function. Instead, I've detached the widget's `drawer.draw` function.

This seems to work (mirror updates when widget is hidden in box, hidden widgets don't display when polled). Only downside is that it's broken my unittest so I'll need to rework that!

Reply:
Awesome!